## Pokemon Stats Webscrape

In [1]:
import pandas as pd
from bs4 import BeautifulSoup
import requests

## Setup

In [2]:
# Pokemon Db Table

url = "https://pokemondb.net/pokedex/all"

response = requests.get(url)

response.status_code

200

In [3]:
soup = BeautifulSoup(response.content, 'html.parser')

In [4]:
results = soup.find('table', {'id': 'pokedex'}).find('tbody')

In [5]:
len(results)

2095

In [6]:
pokemon = results.find_all('tr')

In [7]:
len(pokemon)

1047

In [8]:
pokemon[0]

<tr>
<td class="cell-num cell-fixed" data-sort-value="1"><span class="infocard-cell-img"><span class="img-fixed icon-pkmn" data-alt="Bulbasaur icon" data-src="https://img.pokemondb.net/sprites/sword-shield/icon/bulbasaur.png"></span></span><span class="infocard-cell-data">001</span></td> <td class="cell-name"><a class="ent-name" href="/pokedex/bulbasaur" title="View Pokedex for #001 Bulbasaur">Bulbasaur</a></td><td class="cell-icon"><a class="type-icon type-grass" href="/type/grass">Grass</a><br/> <a class="type-icon type-poison" href="/type/poison">Poison</a></td>
<td class="cell-total">318</td>
<td class="cell-num">45</td>
<td class="cell-num">49</td>
<td class="cell-num">49</td>
<td class="cell-num">65</td>
<td class="cell-num">65</td>
<td class="cell-num">45</td>
</tr>

In [9]:
pokemon[1044]

<tr>
<td class="cell-num cell-fixed" data-sort-value="898"><span class="infocard-cell-img"><span class="img-fixed icon-pkmn" data-alt="Calyrex icon" data-src="https://img.pokemondb.net/sprites/sword-shield/icon/calyrex.png"></span></span><span class="infocard-cell-data">898</span></td> <td class="cell-name"><a class="ent-name" href="/pokedex/calyrex" title="View Pokedex for #898 Calyrex">Calyrex</a></td><td class="cell-icon"><a class="type-icon type-psychic" href="/type/psychic">Psychic</a><br/> <a class="type-icon type-grass" href="/type/grass">Grass</a></td>
<td class="cell-total">500</td>
<td class="cell-num">100</td>
<td class="cell-num">80</td>
<td class="cell-num">80</td>
<td class="cell-num">80</td>
<td class="cell-num">80</td>
<td class="cell-num">80</td>
</tr>

## 1) Obtain Pokemon Data

## Obtain Pokemon Number

In [10]:
pokemon[0].find('span', {'class': "infocard-cell-data"}).get_text()

'001'

In [11]:
pokemon[5].find('td', {'class': "cell-num cell-fixed"}).get_text()

'005'

## Obtain Pokemon Name

In [12]:
pokemon[0].find('a', {'class': "ent-name"}).get_text()

'Bulbasaur'

In [13]:
pokemon[0].find('td', {'class': "cell-name"}).get_text()

'Bulbasaur'

In [14]:
pokemon[3].find('td', {'class': "cell-name"}).get_text()

'Venusaur Mega Venusaur'

In [15]:
pokemon[3].find_all('a')

[<a class="ent-name" href="/pokedex/venusaur" title="View Pokedex for #003 Venusaur">Venusaur</a>,
 <a class="type-icon type-grass" href="/type/grass">Grass</a>,
 <a class="type-icon type-poison" href="/type/poison">Poison</a>]

In [16]:
pokemon[1043].find('td', {'class': "cell-name"}).get_text()

'Spectrier'

## Obtain Pokemon Types

In [17]:
# Venusaur
pokemon[3].find('td', {'class': "cell-icon"}).get_text()

'Grass Poison'

In [18]:
# Charizard
pokemon[6].find('td', {'class': "cell-icon"}).get_text()

'Fire Flying'

In [19]:
# Mega Charizard X
pokemon[7].find('td', {'class': "cell-icon"}).get_text()

'Fire Dragon'

## Total

In [20]:
# Mega Venusaur
pokemon[3].find('td', {'class': "cell-total"}).get_text()

'625'

## HP

In [21]:
# Mega Venusaur
pokemon[3].find_all('td', {'class': "cell-num"})

[<td class="cell-num cell-fixed" data-sort-value="3"><span class="infocard-cell-img"><span class="img-fixed icon-pkmn" data-alt="Mega Venusaur icon" data-src="https://img.pokemondb.net/sprites/sword-shield/icon/venusaur-mega.png"></span></span><span class="infocard-cell-data">003</span></td>,
 <td class="cell-num">80</td>,
 <td class="cell-num">100</td>,
 <td class="cell-num">123</td>,
 <td class="cell-num">122</td>,
 <td class="cell-num">120</td>,
 <td class="cell-num">80</td>]

In [22]:
# Mega Venusaur
pokemon[3].find_all('td', {'class': "cell-num"})[1].get_text()

'80'

## Attack

In [23]:
# Mega Venusaur
pokemon[3].find_all('td', {'class': "cell-num"})[2].get_text()

'100'

## Defense

In [24]:
# Mega Venusaur
pokemon[3].find_all('td', {'class': "cell-num"})[3].get_text()

'123'

## Sp. Attack

In [25]:
# Mega Venusaur
pokemon[3].find_all('td', {'class': "cell-num"})[4].get_text()

'122'

## Sp. Def

In [26]:
# Mega Venusaur
pokemon[3].find_all('td', {'class': "cell-num"})[5].get_text()

'120'

## Speed

In [27]:
# Mega Venusaur
pokemon[3].find_all('td', {'class': "cell-num"})[6].get_text()

'80'

## Make Pokemon Table

In [28]:
# Obtain data into lists (list comprehension):

number = [p.find('td', {'class': "cell-num cell-fixed"}).get_text() for p in pokemon]

name = [p.find('td', {'class': "cell-name"}).get_text() for p in pokemon]

poke_type = [p.find('td', {'class': "cell-icon"}).get_text() for p in pokemon]

total = [p.find('td', {'class': "cell-total"}).get_text() for p in pokemon]

hp = [p.find_all('td', {'class': "cell-num"})[1].get_text() for p in pokemon]

attack = [p.find_all('td', {'class': "cell-num"})[2].get_text() for p in pokemon]

defense = [p.find_all('td', {'class': "cell-num"})[3].get_text() for p in pokemon]

sp_atk = [p.find_all('td', {'class': "cell-num"})[4].get_text() for p in pokemon]

sp_def = [p.find_all('td', {'class': "cell-num"})[5].get_text() for p in pokemon]

speed = [p.find_all('td', {'class': "cell-num"})[6].get_text() for p in pokemon]

In [29]:
## Make pandas Dataframe:

pokemon_df = pd.DataFrame({'Number': number, 'Pokemon': name, 'Type': poke_type,
                           'Total': total, 'HP': hp, 'Attack': attack,
                           'Defense': defense, 'Sp. Atk': sp_atk, 
                           'Sp. Def': sp_def, 'Speed': speed})

In [30]:
pokemon_df.head(12)

Number                     Pokemon          Type Total  HP Attack Defense  \
0     001                   Bulbasaur  Grass Poison   318  45     49      49   
1     002                     Ivysaur  Grass Poison   405  60     62      63   
2     003                    Venusaur  Grass Poison   525  80     82      83   
3     003      Venusaur Mega Venusaur  Grass Poison   625  80    100     123   
4     004                  Charmander         Fire    309  39     52      43   
5     005                  Charmeleon         Fire    405  58     64      58   
6     006                   Charizard   Fire Flying   534  78     84      78   
7     006  Charizard Mega Charizard X   Fire Dragon   634  78    130     111   
8     006  Charizard Mega Charizard Y   Fire Flying   634  78    104      78   
9     007                    Squirtle        Water    314  44     48      65   
10    008                   Wartortle        Water    405  59     63      80   
11    009                   Blastoise        Water    530  79     83     100   

   Sp. Atk Sp. Def Speed  
0       65      65    45  
1       80      80    60  
2      100     100    80  
3      122     120    80  
4       60      50    65  
5       80      65    80  
6      109      85   100  
7      130      85   100  
8      159     115   100  
9       50      64    43  
10      65      80    58  
11      85     105    78

In [31]:
## Check if any NA values

pokemon_df.isna().sum()

Number     0
Pokemon    0
Type       0
Total      0
HP         0
Attack     0
Defense    0
Sp. Atk    0
Sp. Def    0
Speed      0
dtype: int64

## Optional: Save Pokémon Table In .csv File

In [32]:
## Save to .csv File:
pokemon_df.to_csv("Pokédex_db.csv", index = False)

## 2) Data Analysis On Pokémon

Now that the Pokemon table has been extracted, we can do some basic data analysis on the pokemon. We can find the fastest pokemon, the slowest pokemon, pokemon with high attack stats, high defense stats and so on.

In [33]:
# Check Data info:

pokemon_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1047 entries, 0 to 1046
Data columns (total 10 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   Number   1047 non-null   object
 1   Pokemon  1047 non-null   object
 2   Type     1047 non-null   object
 3   Total    1047 non-null   object
 4   HP       1047 non-null   object
 5   Attack   1047 non-null   object
 6   Defense  1047 non-null   object
 7   Sp. Atk  1047 non-null   object
 8   Sp. Def  1047 non-null   object
 9   Speed    1047 non-null   object
dtypes: object(10)
memory usage: 81.9+ KB


In [34]:
## Change to numeric columns on number, total, HP, attack, defense, sp. atk, sp. def and speed

cols = ['Number', 'Total', 'HP', 'Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed']

In [35]:
pokemon_df[cols] = pokemon_df[cols].apply(pd.to_numeric, errors='coerce', axis=1)

### Largest Total Stats Pokemon

In [36]:
# Top 15 Pokemon with highest total stats

pokemon_df.sort_values('Total', ascending = False)[:15]

Number                   Pokemon              Type  Total   HP  Attack  \
1035     890       Eternatus Eternamax     Poison Dragon   1125  255     115   
194      150      Mewtwo Mega Mewtwo X  Psychic Fighting    780  106     190   
195      150      Mewtwo Mega Mewtwo Y          Psychic     780  106     150   
464      384    Rayquaza Mega Rayquaza     Dragon Flying    780  105     180   
460      382      Kyogre Primal Kyogre            Water     770  100     150   
462      383    Groudon Primal Groudon       Ground Fire    770  100     180   
938      800   Necrozma Ultra Necrozma    Psychic Dragon    754   97     167   
1030     888      Zacian Crowned Sword       Fairy Steel    720   92     170   
1032     889  Zamazenta Crowned Shield    Fighting Steel    720   92     130   
592      493                    Arceus           Normal     720  120     120   
843      718    Zygarde Complete Forme     Dragon Ground    708  216     100   
451      376  Metagross Mega Metagross     Steel Psychic    700   80     145   
458      381        Latios Mega Latios    Dragon Psychic    700   80     130   
447      373  Salamence Mega Salamence     Dragon Flying    700   95     145   
845      719      Diancie Mega Diancie        Rock Fairy    700   50     160   

      Defense  Sp. Atk  Sp. Def  Speed  
1035      250      125      250    130  
194       100      154      100    130  
195        70      194      120    140  
464       100      180      100    115  
460        90      180      160     90  
462       160      150       90     90  
938        97      167       97    129  
1030      115       80      115    148  
1032      145       80      145    128  
592       120      120      120    120  
843       121       91       95     85  
451       150      105      110    110  
458       100      160      120    110  
447       130      120       90    120  
845       110      160      110    110

Number one is Eternatus Eternamax with a big total of 1125.

There is a tie for number 2 in total stats. These are Mega Mewtwo X, Mega Mewtwo, Y and Mega Rayquaza with a total stats of 780.

Notable Pokemon in this top 15 list are Mega Garchomp and Mega Metagross.

### Smallest Total Stats Pokemon

In [37]:
# Bottom 15 Pokemon for total stats

pokemon_df.sort_values('Total', ascending = True)[:15]

Number               Pokemon           Type  Total  HP  Attack  Defense  \
879      746  Wishiwashi Solo Form         Water     175  45      20       20   
237      191               Sunkern         Grass     180  30      30       30   
962      824               Blipbug           Bug     180  25      20       20   
1011     872                  Snom        Ice Bug    185  30      25       35   
357      298               Azurill   Normal Fairy    190  50      20       40   
484      401             Kricketot           Bug     194  37      25       41   
323      265               Wurmple           Bug     195  45      45       35   
13        10              Caterpie           Bug     195  45      30       35   
16        13                Weedle     Bug Poison    195  40      35       30   
338      280                 Ralts  Psychic Fairy    198  28      25       25   
416      349                Feebas         Water     200  20      15       20   
924      789                Cosmog       Psychic     200  43      29       31   
166      129              Magikarp         Water     200  20      10       55   
779      664            Scatterbug           Bug     200  38      35       40   
326      268               Cascoon           Bug     205  50      35       55   

      Sp. Atk  Sp. Def  Speed  
879        25       25     40  
237        30       30     30  
962        25       45     45  
1011       45       30     20  
357        20       40     20  
484        25       41     25  
323        20       30     20  
13         20       20     45  
16         20       20     50  
338        45       35     40  
416        10       55     80  
924        29       31     37  
166        15       20     80  
779        27       25     35  
326        25       25     15

Here are the pokemon with the smallest total stats. Number one is **not** Magikarp. It is Wishiwashi Solo Form as number one.

A lot of the bug Pokemon such as Blipbug, Snom, Kricketot, Caterpie, Wurmple and Scatterbug are featured in this list. 

### Fastest Pokemon

In [38]:
# Top 15 Pokemon with highest speed stats

pokemon_df.sort_values('Speed', ascending = False)[:15]

Number                     Pokemon           Type  Total   HP  Attack  \
1040     894                   Regieleki      Electric     580   80     100   
469      386          Deoxys Speed Forme       Psychic     600   50      95   
350      291                     Ninjask     Bug Flying    456   61      90   
930      795                   Pheromosa   Bug Fighting    570   71     137   
1046     898        Calyrex Shadow Rider  Psychic Ghost    680  100      85   
182      142  Aerodactyl Mega Aerodactyl    Rock Flying    615   80     135   
466      386         Deoxys Normal Forme       Psychic     600   50     150   
467      386         Deoxys Attack Forme       Psychic     600   50     180   
84        65      Alakazam Mega Alakazam       Psychic     600   55      50   
132      101                   Electrode      Electric     490   60      50   
1030     888        Zacian Crowned Sword    Fairy Steel    720   92     170   
19        15      Beedrill Mega Beedrill     Bug Poison    495   65     150   
723      617                    Accelgor           Bug     495   80      70   
308      254      Sceptile Mega Sceptile   Grass Dragon    630   70     110   
945      807                     Zeraora      Electric     600   88     112   

      Defense  Sp. Atk  Sp. Def  Speed  
1040       50      100       50    200  
469        90       95       90    180  
350        45       50       50    160  
930        37      137       37    151  
1046       80      165      100    150  
182        85       70       95    150  
466        50      150       50    150  
467        20      180       20    150  
84         65      175      105    150  
132        70       80       80    150  
1030      115       80      115    148  
19         40       15       80    145  
723        40      100       60    145  
308        75      145       85    145  
945        75      102       80    143

### Slowest Pokemon

In [39]:
# Top 15 Pokemon with low speed stats

pokemon_df.sort_values('Speed', ascending = True)[:15]

Number     Pokemon          Type  Total   HP  Attack  Defense  Sp. Atk  \
535     446    Munchlax       Normal     390  135      85       40       40   
905     771   Pyukumuku        Water     410   55      60      130       30   
262     213     Shuckle      Bug Rock    505   20      10      230       10   
703     597   Ferroseed   Grass Steel    305   44      50       91       24   
394     328    Trapinch       Ground     290   45     100       45       45   
526     438      Bonsly         Rock     290   50      80       95       10   
967     829  Gossifleur        Grass     250   40      40       60       40   
943     805   Stakataka    Rock Steel    570   61     131      211       53   
903     769   Sandygast  Ghost Ground    320   55      55       80       70   
889     755    Morelull   Grass Fairy    285   40      35       55       65   
623     524  Roggenrola         Rock     280   55      75       85       25   
218     173      Cleffa        Fairy     218   50      25       28       45   
219     174   Igglybuff  Normal Fairy    210   90      30       15       40   
324     266     Silcoon          Bug     205   50      35       55       25   
251     204      Pineco          Bug     290   50      65       90       35   

     Sp. Def  Speed  
535       85      5  
905      130      5  
262      230      5  
703       86     10  
394       45     10  
526       45     10  
967       60     10  
943      101     13  
903       45     15  
889       75     15  
623       25     15  
218       55     15  
219       20     15  
324       25     15  
251       35     15

### Highest Attack Pokemon

In [40]:
# Top 15 Pokemon with highest attack stats

pokemon_df.sort_values('Attack', ascending = False)[:15]

Number                   Pokemon              Type  Total   HP  Attack  \
194      150      Mewtwo Mega Mewtwo X  Psychic Fighting    780  106     190   
264      214  Heracross Mega Heracross      Bug Fighting    600   80     185   
933      798                   Kartana       Grass Steel    570   59     181   
464      384    Rayquaza Mega Rayquaza     Dragon Flying    780  105     180   
462      383    Groudon Primal Groudon       Ground Fire    770  100     180   
467      386       Deoxys Attack Forme          Psychic     600   50     180   
757      646       Kyurem Black Kyurem        Dragon Ice    700  125     170   
534      445    Garchomp Mega Garchomp     Dragon Ground    700  108     170   
1030     888      Zacian Crowned Sword       Fairy Steel    720   92     170   
938      800   Necrozma Ultra Necrozma    Psychic Dragon    754   97     167   
425      354      Banette Mega Banette            Ghost     555   64     165   
492      409                 Rampardos             Rock     495   97     165   
567      475      Gallade Mega Gallade  Psychic Fighting    618   68     165   
1045     898         Calyrex Ice Rider       Psychic Ice    680  100     165   
301      248  Tyranitar Mega Tyranitar         Rock Dark    700  100     164   

      Defense  Sp. Atk  Sp. Def  Speed  
194       100      154      100    130  
264       115       40      105     75  
933       131       59       31    109  
464       100      180      100    115  
462       160      150       90     90  
467        20      180       20    150  
757       100      120       90     95  
534       115      120       95     92  
1030      115       80      115    148  
938        97      167       97    129  
425        75       93       83     75  
492        60       65       50     58  
567        95       65      115    110  
1045      150       85      130     50  
301       150       95      120     71

### Lowest Attack Pokemon

In [41]:
# Top 15 Pokemon with lowest attack stats

pokemon_df.sort_values('Attack', ascending = True)[:15]

Number               Pokemon          Type  Total   HP  Attack  Defense  \
528     440               Happiny       Normal     220  100       5        5   
147     113               Chansey       Normal     450  250       5        5   
262     213               Shuckle      Bug Rock    505   20      10      230   
294     242               Blissey       Normal     540  255      10       10   
166     129              Magikarp        Water     200   20      10       55   
416     349                Feebas        Water     200   20      15       20   
357     298               Azurill  Normal Fairy    190   50      20       40   
220     175                Togepi        Fairy     245   35      20       65   
229     183                Marill   Water Fairy    250   70      20       50   
962     824               Blipbug          Bug     180   25      20       20   
879     746  Wishiwashi Solo Form        Water     175   45      20       20   
81       63                  Abra      Psychic     310   25      20       15   
548     458               Mantyke  Water Flying    345   45      20       50   
210     165                Ledyba    Bug Flying    265   40      20       30   
287     235              Smeargle       Normal     250   55      20       35   

     Sp. Atk  Sp. Def  Speed  
528       15       65     30  
147       35      105     50  
262       10      230      5  
294       75      135     55  
166       15       20     80  
416       10       55     80  
357       20       40     20  
220       40       65     20  
229       20       50     40  
962       25       45     45  
879       25       25     40  
81       105       55     90  
548       60      120     50  
210       40       80     55  
287       20       45     75

### Highest & Lowest Special Attack Pokemon

In [42]:
# Top 15 Pokemon with highest special attack stats

pokemon_df.sort_values('Sp. Atk', ascending = False)[:15]

Number                   Pokemon             Type  Total   HP  Attack  \
195      150      Mewtwo Mega Mewtwo Y         Psychic     780  106     150   
464      384    Rayquaza Mega Rayquaza    Dragon Flying    780  105     180   
467      386       Deoxys Attack Forme         Psychic     600   50     180   
460      382      Kyogre Primal Kyogre           Water     770  100     150   
84        65    Alakazam Mega Alakazam         Psychic     600   55      50   
931      796                 Xurkitree        Electric     570   83      89   
125       94        Gengar Mega Gengar     Ghost Poison    600   60      65   
847      720       Hoopa Hoopa Unbound     Psychic Dark    680   80     160   
758      646       Kyurem White Kyurem       Dragon Ice    700  125     120   
938      800   Necrozma Ultra Necrozma   Psychic Dragon    754   97     167   
341      282  Gardevoir Mega Gardevoir    Psychic Fairy    618   68      85   
1046     898      Calyrex Shadow Rider    Psychic Ghost    680  100      85   
227      181    Ampharos Mega Ampharos  Electric Dragon    610   90      95   
845      719      Diancie Mega Diancie       Rock Fairy    700   50     160   
458      381        Latios Mega Latios   Dragon Psychic    700   80     130   

      Defense  Sp. Atk  Sp. Def  Speed  
195        70      194      120    140  
464       100      180      100    115  
467        20      180       20    150  
460        90      180      160     90  
84         65      175      105    150  
931        71      173       71     83  
125        80      170       95    130  
847        60      170      130     80  
758        90      170      100     95  
938        97      167       97    129  
341        65      165      135    100  
1046       80      165      100    150  
227       105      165      110     45  
845       110      160      110    110  
458       100      160      120    110

In [43]:
# Top 15 Pokemon with highest special attack stats

pokemon_df.sort_values('Sp. Atk', ascending = True)[:15]

Number                     Pokemon         Type  Total   HP  Attack  \
526      438                      Bonsly        Rock     290   50      80   
37        27  Sandshrew Alolan Sandshrew    Ice Steel    300   50      75   
262      213                     Shuckle     Bug Rock    505   20      10   
416      349                      Feebas       Water     200   20      15   
656      554  Darumaka Galarian Darumaka         Ice     315   70      90   
655      554                    Darumaka        Fire     315   70      90   
528      440                     Happiny      Normal     220  100       5   
166      129                    Magikarp       Water     200   20      10   
19        15      Beedrill Mega Beedrill   Bug Poison    495   65     150   
229      183                      Marill  Water Fairy    250   70      20   
901      767                      Wimpod    Bug Water    230   25      35   
287      235                    Smeargle      Normal     250   55      20   
1013     874                 Stonjourner        Rock     470  100     125   
16        13                      Weedle   Bug Poison    195   40      35   
36        27                   Sandshrew      Ground     300   50      75   

      Defense  Sp. Atk  Sp. Def  Speed  
526        95       10       45     10  
37         90       10       35     40  
262       230       10      230      5  
416        20       10       55     80  
656        45       15       45     50  
655        45       15       45     50  
528         5       15       65     30  
166        55       15       20     80  
19         40       15       80    145  
229        50       20       50     40  
901        40       20       30     80  
287        35       20       45     75  
1013      135       20       20     70  
16         30       20       20     50  
36         85       20       30     40

### Highest & Lowest Defense Pokemon

In [44]:
# Top 15 Pokemon with highest defense stats

pokemon_df.sort_values('Defense', ascending = False)[:15]

Number                 Pokemon           Type  Total   HP  Attack  \
1035     890     Eternatus Eternamax  Poison Dragon   1125  255     115   
256      208    Steelix Mega Steelix   Steel Ground    610   75     125   
368      306      Aggron Mega Aggron         Steel     630   70     140   
262      213                 Shuckle       Bug Rock    505   20      10   
943      805               Stakataka     Rock Steel    570   61     131   
255      208                 Steelix   Steel Ground    510   75      85   
452      377                Regirock          Rock     580   80     100   
836      713                 Avalugg           Ice     514   95     117   
367      306                  Aggron     Steel Rock    530   70     110   
106       80    Slowbro Mega Slowbro  Water Psychic    590   95      75   
121       91                Cloyster      Water Ice    525   50      95   
494      411               Bastiodon     Rock Steel    495   60      52   
468      386    Deoxys Defense Forme       Psychic     600   50      70   
126       95                    Onix    Rock Ground    385   35      45   
462      383  Groudon Primal Groudon    Ground Fire    770  100     180   

      Defense  Sp. Atk  Sp. Def  Speed  
1035      250      125      250    130  
256       230       55       95     30  
368       230       60       80     50  
262       230       10      230      5  
943       211       53      101     13  
255       200       55       65     30  
452       200       50      100     50  
836       184       44       46     28  
367       180       60       60     50  
106       180      130       80     30  
121       180       85       45     70  
494       168       47      138     30  
468       160       70      160     90  
126       160       30       45     70  
462       160      150       90     90

In [45]:
# Top 15 Pokemon with lowest defense stats

pokemon_df.sort_values('Defense', ascending = True)[:15]

Number               Pokemon          Type  Total   HP  Attack  Defense  \
528     440               Happiny       Normal     220  100       5        5   
147     113               Chansey       Normal     450  250       5        5   
294     242               Blissey       Normal     540  255      10       10   
217     172                 Pichu     Electric     205   20      40       15   
81       63                  Abra      Psychic     310   25      20       15   
290     238              Smoochum   Ice Psychic    305   45      30       15   
219     174             Igglybuff  Normal Fairy    210   90      30       15   
382     318              Carvanha    Water Dark    305   45      90       20   
962     824               Blipbug          Bug     180   25      20       20   
52       39            Jigglypuff  Normal Fairy    270  115      45       20   
467     386   Deoxys Attack Forme      Psychic     600   50     180       20   
416     349                Feebas        Water     200   20      15       20   
879     746  Wishiwashi Solo Form        Water     175   45      20       20   
352     293               Whismur       Normal     240   64      51       23   
63       50               Diglett       Ground     265   10      55       25   

     Sp. Atk  Sp. Def  Speed  
528       15       65     30  
147       35      105     50  
294       75      135     55  
217       35       35     60  
81       105       55     90  
290       85       65     65  
219       40       20     15  
382       65       20     65  
962       25       45     45  
52        45       25     20  
467      180       20    150  
416       10       55     80  
879       25       25     40  
352       51       23     28  
63        35       45     95

### Highest & Lowest Special Defense Pokemon

In [46]:
# Top 15 Pokemon with highest special defense stats

pokemon_df.sort_values('Sp. Def', ascending = False)[:15]

Number                   Pokemon            Type  Total   HP  Attack  \
1035     890       Eternatus Eternamax   Poison Dragon   1125  255     115   
262      213                   Shuckle        Bug Rock    505   20      10   
453      378                    Regice            Ice     580   80      50   
460      382      Kyogre Primal Kyogre          Water     770  100     150   
468      386      Deoxys Defense Forme        Psychic     600   50      70   
786      671                   Florges          Fairy     552   78      65   
302      249                     Lugia  Psychic Flying    680  106      90   
303      250                     Ho-oh     Fire Flying    680  106     130   
823      706                    Goodra         Dragon     600   90     100   
456      380        Latias Mega Latias  Dragon Psychic    700   80     100   
844      719                   Diancie      Rock Fairy    600   50     100   
568      476                 Probopass      Rock Steel    525   60      55   
820      703                   Carbink      Rock Fairy    500   50      50   
454      379                 Registeel          Steel     580   80      75   
1032     889  Zamazenta Crowned Shield  Fighting Steel    720   92     130   

      Defense  Sp. Atk  Sp. Def  Speed  
1035      250      125      250    130  
262       230       10      230      5  
453       100      100      200     50  
460        90      180      160     90  
468       160       70      160     90  
786        68      112      154     75  
302       130       90      154    110  
303        90      110      154     90  
823        70      110      150     80  
456       120      140      150    110  
844       150      100      150     50  
568       145       75      150     40  
820       150       50      150     50  
454       150       75      150     50  
1032      145       80      145    128

In [47]:
# Top 15 Pokemon with highest special defense stats

pokemon_df.sort_values('Sp. Def', ascending = True)[:15]

Number              Pokemon          Type  Total   HP  Attack  Defense  \
1013     874          Stonjourner         Rock     470  100     125      135   
467      386  Deoxys Attack Forme      Psychic     600   50     180       20   
219      174            Igglybuff  Normal Fairy    210   90      30       15   
382      318             Carvanha    Water Dark    305   45      90       20   
166      129             Magikarp        Water     200   20      10       55   
16        13               Weedle    Bug Poison    195   40      35       30   
13        10             Caterpie          Bug     195   45      30       35   
352      293              Whismur       Normal     240   64      51       23   
120       90             Shellder        Water     305   30      65      100   
52        39           Jigglypuff  Normal Fairy    270  115      45       20   
17        14               Kakuna    Bug Poison    205   45      25       50   
151      116               Horsea        Water     295   30      40       70   
14        11              Metapod          Bug     205   50      20       55   
326      268              Cascoon          Bug     205   50      35       55   
324      266              Silcoon          Bug     205   50      35       55   

      Sp. Atk  Sp. Def  Speed  
1013       20       20     70  
467       180       20    150  
219        40       20     15  
382        65       20     65  
166        15       20     80  
16         20       20     50  
13         20       20     45  
352        51       23     28  
120        45       25     40  
52         45       25     20  
17         25       25     35  
151        70       25     60  
14         25       25     30  
326        25       25     15  
324        25       25     15

## Mew & Mewtwo

In [48]:
pokemon_df[pokemon_df['Pokemon'].str.contains('Mew')]

Number               Pokemon              Type  Total   HP  Attack  \
193     150                Mewtwo          Psychic     680  106     110   
194     150  Mewtwo Mega Mewtwo X  Psychic Fighting    780  106     190   
195     150  Mewtwo Mega Mewtwo Y          Psychic     780  106     150   
196     151                   Mew          Psychic     600  100     100   

     Defense  Sp. Atk  Sp. Def  Speed  
193       90      154       90    130  
194      100      154      100    130  
195       70      194      120    140  
196      100      100      100    100

## Pseudo Legendary Pokemon Stats

Link: https://bulbapedia.bulbagarden.net/wiki/Pseudo-legendary_Pok%C3%A9mon

Pseudo Legendary Pokemon

* Dragonite (Dragon/Flying)
* Tyranitar (Rock/Dark)
* Salamence (Dragon/Flying)
* Metagross (Steel/Psychic)
* Garchomp (Dragon/Ground)
* Hydreigon (Dark/Dragon)
* Goodra (Dragon)
* Kommo-o (Dragon/Fighting)
* Dragapult (Dragon/Ghost)

In [49]:
pseudo_legends = ['Dragonite', 'Tyranitar', 'Salamence', 'Metagross', 'Garchomp',
                  'Hydreigon', 'Goodra', 'Kommo-o', 'Dragapult']

In [50]:
# Filter pokemon_df based on pokemon name being in pseudo legendaries list

pokemon_df[pokemon_df['Pokemon'].isin(pseudo_legends)]

Number    Pokemon             Type  Total   HP  Attack  Defense  \
192      149  Dragonite    Dragon Flying    600   91     134       95   
300      248  Tyranitar        Rock Dark    600  100     134      110   
446      373  Salamence    Dragon Flying    600   95     135       80   
450      376  Metagross    Steel Psychic    600   80     135      130   
533      445   Garchomp    Dragon Ground    600  108     130       95   
742      635  Hydreigon      Dark Dragon    600   92     105       90   
823      706     Goodra          Dragon     600   90     100       70   
919      784    Kommo-o  Dragon Fighting    600   75     110      125   
1029     887  Dragapult     Dragon Ghost    600   88     120       75   

      Sp. Atk  Sp. Def  Speed  
192       100      100     80  
300        95      100     61  
446       110       80    100  
450        95       90     70  
533        80       85    102  
742       125       90     98  
823       110      150     80  
919       100      105     85  
1029      100       75    142

All of these pseudo legendary pokemon have a total stats of 600.

## Bug Types Top 10

In [51]:
# Filter pokemon_df based on Fairy type

pokemon_df[pokemon_df['Type'].str.contains('Bug')].sort_values('Total', ascending = False)[:10]

Number                   Pokemon          Type  Total   HP  Attack  \
164     127        Pinsir Mega Pinsir    Bug Flying    600   65     155   
763     649                  Genesect     Bug Steel    600   71     120   
261     212        Scizor Mega Scizor     Bug Steel    600   70     150   
264     214  Heracross Mega Heracross  Bug Fighting    600   80     185   
930     795                 Pheromosa  Bug Fighting    570   71     137   
929     794                  Buzzwole  Bug Fighting    570  107     139   
744     637                 Volcarona      Bug Fire    550   85      60   
902     768                 Golisopod     Bug Water    530   75     125   
990     851               Centiskorch      Fire Bug    525  100     115   
560     469                   Yanmega    Bug Flying    515   86      76   

     Defense  Sp. Atk  Sp. Def  Speed  
164      120       65       90    105  
763       95      120       95     99  
261      140       65      100     75  
264      115       40      105     75  
930       37      137       37    151  
929      139       53       53     79  
744       65      135      105    100  
902      140       60       90     40  
990       65       90       90     65  
560       86      116       56     95

## Fairy Types Top 10

In [52]:
# Filter pokemon_df based on Fairy type

pokemon_df[pokemon_df['Type'].str.contains('Fairy')].sort_values('Total', ascending = False)[:10]

Number                      Pokemon           Type  Total   HP  Attack  \
1030     888         Zacian Crowned Sword    Fairy Steel    720   92     170   
845      719         Diancie Mega Diancie     Rock Fairy    700   50     160   
839      716                      Xerneas         Fairy     680  126     131   
1031     888  Zacian Hero of Many Battles         Fairy     670   92     130   
341      282     Gardevoir Mega Gardevoir  Psychic Fairy    618   68      85   
939      801                     Magearna    Steel Fairy    600   80      95   
844      719                      Diancie     Rock Fairy    600   50     100   
401      334         Altaria Mega Altaria   Dragon Fairy    590   75     110   
923      788                    Tapu Fini    Water Fairy    570   70      75   
922      787                    Tapu Bulu    Grass Fairy    570   70     130   

      Defense  Sp. Atk  Sp. Def  Speed  
1030      115       80      115    148  
845       110      160      110    110  
839        95      131       98     99  
1031      115       80      115    138  
341        65      165      135    100  
939       115      130      115     65  
844       150      100      150     50  
401       110      110      105     80  
923       115       95      130     85  
922       115       85       95     75

## Water Types Top 10

In [53]:
# Filter pokemon_df based on Fairy type

pokemon_df[pokemon_df['Type'].str.contains('Water')].sort_values('Total', ascending = False)[:10]

Number                   Pokemon          Type  Total   HP  Attack  \
460     382      Kyogre Primal Kyogre        Water     770  100     150   
581     484                    Palkia  Water Dragon    680   90     120   
459     382                    Kyogre        Water     670  100     100   
773     658     Greninja Ash-Greninja    Water Dark    640   72     145   
168     130    Gyarados Mega Gyarados    Water Dark    640   95     155   
316     260    Swampert Mega Swampert  Water Ground    635  100     150   
12        9  Blastoise Mega Blastoise        Water     630   79     103   
880     746    Wishiwashi School Form        Water     620   45     140   
588     490                   Manaphy        Water     600  100     100   
848     721                 Volcanion    Fire Water    600   80     110   

     Defense  Sp. Atk  Sp. Def  Speed  
460       90      180      160     90  
581      100      150      120    100  
459       90      150      140     90  
773       67      153       71    132  
168      109       70      130     81  
316      110       95      110     70  
12       120      135      115     78  
880      130      140      135     30  
588      100      100      100    100  
848      120      130       90     70